Downward API implementation for resources limits and requests

aveshagarwal opened the pull request:
This is an implementation of Downward API for resources limits and requests, and it works with environment variables and volume plugin.

This is based on proposal https://github.com/kubernetes/kubernetes/pull/24051. This implementation follows API with magic keys approach as discussed in the proposal.

@kubernetes/rh-cluster-infra

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/24179)

<!-- Reviewable:end -->

aveshagarwal commented:
Updated  PR. Will add more tests soon.
@bgrant0607 @derekwaynecarr @pmorie PTAL.

aveshagarwal commented:
Added some validation tests and fixed some issues.

aveshagarwal commented:
I have tested it with following in case any one is interested:

Env variables:

```
apiVersion: v1
kind: Pod
metadata:
  name: dapi-test-pod
spec:
  containers:
    - name: test-container
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh", "-c", "env" ]
      resources:
        requests:
          memory: "64Mi"
          cpu: "250m"
        limits:
          memory: "128Mi"
          cpu: "1500m"
      env:
        - name: CPU_LIMIT
          valueFrom:
            resourceFieldRef:
              resource: limits.cpu
    - name: test-container2
      image: gcr.io/google_containers/busybox
      command: [ "/bin/sh", "-c", "env" ]
      resources:
        requests:
          memory: "32Mi"
          cpu: "125m"
        limits:
          memory: "64Mi"
          cpu: "250m"
      env:
        - name: MEMORY_LIMIT
          valueFrom:
            resourceFieldRef:
              resource: limits.memory
        - name: CPU_LIMIT
          valueFrom:
            resourceFieldRef:
              resource: requests.cpu
  restartPolicy: Never
```

Volume plugin:

```
apiVersion: v1
kind: Pod
metadata:
  name: kubernetes-downwardapi-volume-example
  labels:
    zone: us-est-coast
    cluster: test-cluster1
    rack: rack-22
  annotations:
    build: two
    builder: john-doe
spec:
  containers:
    - name: client-container
      image: gcr.io/google_containers/busybox
      command: ["sh", "-c", "while true; do if [[ -e /etc/labels ]]; then cat /etc/labels; fi; if [[ -e /etc/annotations ]]; then cat /etc/annotations; fi; sleep 5; done"]
      resources:
        requests:
          memory: "32Mi"
          cpu: "125m"
        limits:
          memory: "64Mi"
          cpu: "250m"
      volumeMounts:
        - name: podinfo
          mountPath: /etc
          readOnly: false
  volumes:
    - name: podinfo
      downwardAPI:
        items:
          - path: "labels"
            fieldRef:
              fieldPath: metadata.labels
          - path: "annotations"
            fieldRef:
              fieldPath: metadata.annotations
          - path: "client-container/pod_name"
            fieldRef:
              fieldPath: metadata.name
          - path: "cpu_limit"
            resourceFieldRef:
              containerName: client-container
              resource: limits.cpu
              divisor: "1"
          - path: "cpu_request"
            resourceFieldRef:
              containerName: client-container
              resource: requests.cpu
              divisor: "1m"
          - path: "memory_limit"
            resourceFieldRef:
              containerName: client-container
              resource: limits.memory
              divisor: "1M"
          - path: "memory_request"
            resourceFieldRef:
              containerName: client-container
              resource: requests.memory
```

aveshagarwal commented:
@bgrant0607 @derekwaynecarr @pmorie I have tested it with the above configs and also some other variations including error conditions.

pmorie commented on pkg/api/types.go:
```diff
@@ -827,6 +833,16 @@ type ObjectFieldSelector struct {
+	ContainerName string `json:"containerName,omitempty"`
+	// Required: resource to select
+	Resource string `json:"resource"`
+	// Specifies the output format of the exposed resources
```
Add info about the default here.

pmorie commented on pkg/api/types.go:
```diff
@@ -827,6 +833,16 @@ type ObjectFieldSelector struct {
+	// Container name
+	ContainerName string `json:"containerName,omitempty"`
+	// Required: resource to select
+	Resource string `json:"resource"`
```
Shouldn't this be a `ResourceName`?  cc @derekwaynecarr

pmorie replied:
My bad, this is okay as a string, but there should be a note saying which expressions are supported.

aveshagarwal replied:
Actually that note is there with type ResourceFieldRef.

pmorie commented on pkg/api/validation/validation.go:
```diff
@@ -702,7 +702,16 @@ func validateDownwardAPIVolumeSource(downwardAPIVolume *api.DownwardAPIVolumeSou
+		} else if downwardAPIVolumeFile.ResourceFieldRef != nil {
+			allErrs = append(allErrs, validateContainerResourceFieldSelector(downwardAPIVolumeFile.ResourceFieldRef, &validContainerResourceFieldPathExpressions, fldPath.Child("resourceFieldRef"), true)...)
+		} else {
+			allErrs = append(allErrs, field.Required(fldPath.Child("path"), "one of fieldRef and resourceFieldRef is required"))
```
this doesn't have to do with the `path` element, just pass `fldPath` instead.

pmorie commented on pkg/api/validation/validation.go:
```diff
@@ -1101,6 +1115,43 @@ func validateObjectFieldSelector(fs *api.ObjectFieldSelector, expressions *sets.
+		allErrs = append(allErrs, field.Required(fldPath.Child("containerName"), ""))
+	} else if len(fs.Resource) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("resource"), ""))
+	} else if !expressions.Has(fs.Resource) {
```
I would just reference the constant inline instead of making it a parameter to this method.

aveshagarwal replied:
But thats how it is done for other pod level resources like metadata.name etc.

pmorie replied:
accepted, OK to leave as is.

pmorie commented on pkg/api/validation/validation.go:
```diff
@@ -1101,6 +1115,43 @@ func validateObjectFieldSelector(fs *api.ObjectFieldSelector, expressions *sets.
+	if unsetDivisor.Cmp(divisor) == 0 {
+		return allErrs
+	}
+	cpuMilliCoreDivisor := resource.MustParse("1m")
```
aren't these constants?

aveshagarwal replied:
yes.

pmorie commented on pkg/api/types.go:
```diff
@@ -827,6 +833,16 @@ type ObjectFieldSelector struct {
 
+// ResourceFieldSelector represents container resources (cpu, memory) and their output format
+type ResourceFieldSelector struct {
+	// Container name
```
Add a note saying this is required if used in the context of a volume, optional for container.

pmorie commented on pkg/kubectl/describe.go:
```diff
@@ -895,6 +900,9 @@ func describeContainers(label string, containers []api.Container, containerStatu
 					valueFrom = resolverFn(e)
 				}
 				fmt.Fprintf(out, "      %s:\t%s (%s:%s)\n", e.Name, valueFrom, e.ValueFrom.FieldRef.APIVersion, e.ValueFrom.FieldRef.FieldPath)
+			/*case e.ValueFrom.ResourceFieldRef != nil:
```
TODO: finish :)

aveshagarwal replied:
Nice catch. I forgot I will fix it, i think it requires to pass pod to the function.

pmorie commented:
@aveshagarwal this is moving in the right direction.  Please add e2es for envs and volumes.

Commit:
Downward API implementation for resources limits and requests

aveshagarwal commented:
@bgrant0607 @derekwaynecarr @pmorie  I have updated this PR. 

1. Added e2e tests for env vars, and tested locally and they work.
2. Improved validation for memory divisor.
3. Addressed pmorie's comments.

Still working on e2e for volume.

aveshagarwal commented:
Added e2e tests for volumes too now and tested locally to make sure they work. 

@bgrant0607 @pmorie @derekwaynecarr PTAL.

pmorie commented on pkg/api/validation/validation.go:
```diff
@@ -702,7 +702,16 @@ func validateDownwardAPIVolumeSource(downwardAPIVolume *api.DownwardAPIVolumeSou
+		if downwardAPIVolumeFile.FieldRef != nil {
+			allErrs = append(allErrs, validateObjectFieldSelector(downwardAPIVolumeFile.FieldRef, &validDownwardAPIFieldPathExpressions, fldPath.Child("fieldRef"))...)
+			if downwardAPIVolumeFile.ResourceFieldRef != nil {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("path"), "resource", "fieldRef and resourceFieldRef can not be specified simultaneously"))
```
s/\.Child\("path")//

@aveshagarwal

pmorie commented on pkg/api/validation/validation.go:
```diff
@@ -1101,6 +1115,42 @@ func validateObjectFieldSelector(fs *api.ObjectFieldSelector, expressions *sets.
 	return allErrs
 }
 
+func validateContainerResourceFieldSelector(fs *api.ResourceFieldSelector, expressions *sets.String, fldPath *field.Path, volumeDapi bool) field.ErrorList {
```
s/volumeDapi/volume/

pmorie commented:
@aveshagarwal I'll take the nits in a follow-up.  LGTM.  Thanks a lot.

aveshagarwal commented:
I see a failure in unit tests: FAIL	k8s.io/kubernetes/pkg/api , it seems to be failing in pkg/api/serialization_test.go.

aveshagarwal commented:
I regenerated auto-generated files and when I test ./hack/test-go.sh pkg/api now it fails in pkg/api/copy_test.go.  Not sure what is happening.

aveshagarwal commented:
I have found the solution and will just run the tests locally and update it and hopefully it should be good to go.

pmorie commented:
I will re-tag once green

aveshagarwal commented:
This e2e failure does not seem related to this PR:

```
[Fail] [k8s.io] Pods [It] should invoke init containers on a RestartNever pod 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pods.go:710
```

aveshagarwal commented:
Seems similar to this: https://github.com/kubernetes/kubernetes/issues/26131

pmorie commented:
@k8s-bot test this issue #26131

aveshagarwal commented:
@pmorie all green, LGTM again?

aveshagarwal commented:
@bgrant0607  @derekwaynecarr can you tag it LGTM again?
